web/src/main/java/org/springframework/security/web/savedrequest/CookieRequestCache.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.web.savedrequest;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.web.util.UrlUtils;
+import org.springframework.security.web.util.matcher.AnyRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+import org.springframework.web.util.WebUtils;
+
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+
+/**
+ * {@code RequestCache} which stores the {@code SavedRequest} in a cookie.
+ *
+ * @author Zeeshan Adnan
+ */
+public class CookieRequestCache implements RequestCache {
+
+	private RequestMatcher requestMatcher = AnyRequestMatcher.INSTANCE;
+	protected final Log logger = LogFactory.getLog(this.getClass());
+
+	private String cookieName = "saved_request";
+	private String cookiePath = null;
+	private int cookieMaxAge = -1;
+
+	@Override
+	public void saveRequest(HttpServletRequest request, HttpServletResponse response) {
+		if (requestMatcher.matches(request)) {
+			String redirectUrl = UrlUtils.buildFullRequestUrl(request);
+			Cookie savedCookie = new Cookie(cookieName, redirectUrl);
+			savedCookie.setMaxAge(cookieMaxAge);
+			savedCookie.setSecure(request.isSecure());
+			savedCookie.setPath(cookiePath);
+			savedCookie.setHttpOnly(true);
+			response.addCookie(savedCookie);
+		} else {
+			logger.debug("Request not saved as configured RequestMatcher did not match");
+		}
+	}
+
+	@Override
+	public SavedRequest getRequest(HttpServletRequest request, HttpServletResponse response) {
+		Cookie savedRequestCookie = WebUtils.getCookie(request, cookieName);
+		return savedRequestCookie != null ? new SimpleSavedRequest(savedRequestCookie.getValue()) : null;
```

Comment:
This should use `DefaultSavedRequest` instead of `SimpleSavedRequest`

Reply:
@eleftherias  i did not find a suitable constructor to do it in an easy way. Also in the `getMatchingRequest` method its always going to return null. Is it going to make a difference or i am missing something? Any suggestion on a better approach will be very helpful. Thanks.

Reply:
@zeeshanadnan It can use the builder, 
```
new DefaultSavedRequest.Builder().setRequestURI(savedRequestCookie.getValue()).build()
```